### ISSUE TITLE
CTC implementation?

### ISSUE THREAD

**Issue by DabiaoMa** [Microsoft/CNTK#1992]
I have tried several days to understand the methods ' labels_to_graph',  'forward_backward' and 'edit_distance_error' (python API),  which are claimed to be implementation of CTC.  But I failed due to my lack of knowledge.

This is my code to test the three methods
"  
   import cntk as C
   import numpy as np
   input = C.sequence.input_variable((2))
   f = C.layers.Dense(3)
   output = f(input)
   labels = C.sequence.input_variable((3))
   graph = C.labels_to_graph(labels)
   cr = C.forward_backward(graph, output, 2)
   err = C.edit_distance_error(labels, output, 0, 1, 1, True, [2])

   x  = np.array([[1, 2], [3, 4], [5, 6], [7, 8]], dtype=np.float32)
   y = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1], [1, 0 ,0]], dtype=np.float32)
"
The first question is that what the dimensions of  'y' stand for ?  time series and labels one-hot?

IF the first dimension of 'x' and 'y' stands for time series, then I evaluated the 'err' using " err.eval({input : x, labels : y}) ", I can get an answer " array(2.6666667461395264, dtype=float32) ". But if I change the shape of 'x' from (4,2) to (3,2), it becomes invalid, which means the time series of 'labels' and 'output' should be the same. But CTC is supposed to solve the problem that #labels_sequences and #output_sequences are different and vary for different training examples, right?

I then built a simple trainer to train the parameters with 'cr' being the criterion.

"
   lr_schedule = C.learning_rate_schedule(0.1, C.UnitType.minibatch)
   learner = C.sgd(output.parameters, lr_schedule)
   trainer = C.Trainer(output, cr, learner)
   for i in xrange(100):
       trainer.train_minibatch({input : x, labels : y})
   cr.eval({input : x, labels : y})
   output.eval({input : x, labels : y})

"
It turns out that 'cr' indeed becomes very small, but the output is "
 [array([[ -0.71000862,  -0.92857784,   3.09745359],
       [ -1.09405029,  -1.10089934,   7.38608551],
       [ -1.47809196,  -1.27322078,  11.67471886],
       [ -1.86213362,  -1.4455421 ,  15.9633503 ]], dtype=float32)]", 
Does it mean that most possible output sequence is "blank blank blank blank"??

Those questions are driving me mad. Please be so kind to give me some hints if you are aware of the mistakes I have made.

Regards,

**Comment by DabiaoMa** [comment]
I guess cntk does not really implement CTC for python API.... Now I have to go for other DL frameworks....

**Comment by vmazalov** [comment]
Here we go https://github.com/Microsoft/CNTK/blob/master/Tutorials/CNTK_208_Speech_Connectionist_Temporal_Classification.ipynb

**Comment by delzac** [comment]
For anyone looking for a CTC Encoder to convert your labels into the ctc data format that cntk will accept you can check out my extension library [cntkx](https://github.com/delzac/cntkx), it has a `CTCEncoder` to help do just that.
